x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -520,6 +534,78 @@ export class SessionIndex {
     }
   }
 
+  /**
+   * Checks whether specific session is within a concurrent sessions limit.
+   * @param sessionValue Session index value to check against concurrent sessions limit.
+   */
+  async isWithinConcurrentSessionLimit(sessionValue: Readonly<SessionIndexValue>) {
+    const maxConcurrentSessions = this.options.config.session.concurrentSessions?.maxSessions;
+    if (maxConcurrentSessions == null || !sessionValue.usernameHash) {
+      return true;
+    }
+
+    let sessionsOutsideLimit: Array<SearchHit<SessionIndexValue>>;
+    try {
+      const searchResponse = await this.options.elasticsearchClient.search<SessionIndexValue>({
+        index: this.aliasName,
+
+        // Find all sessions created for the same user by the same authentication provider.
+        query: {
+          bool: {
+            must: [
+              { term: { usernameHash: sessionValue.usernameHash } },
+              { term: { 'provider.type': sessionValue.provider.type } },
+              { term: { 'provider.name': sessionValue.provider.name } },
+            ],
+          },
+        },
```

Comment:
> Why differentiate by authentication provider? Shouldn't access be restricted on a per user basis?

Yeah, that might seem confusing, but that's how we historically determined unique users since `username` might be the same for completely different users in different security realms\providers of the same or different types (e.g. `admin` username in two unrelated SAML\LDAP realms\providers).

The only correct way to "join" users from different realms is to rely on [security domains](https://www.elastic.co/guide/en/elasticsearch/reference/master/security-domain.html) (username + security domain instead of username + provider type + provider name __if__ security domains are configured), but since we don't use them widely yet, they are still optional and we don't record domain in the session document we decided (cc @arisonl) to go with the simplest option for MVP and gather feedback.

Does that make sense?

Reply:
Yeh, given those restrictions I think that makes sense.